# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](tttboard.jpg)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [58]:
# your code here
import pandas as pd

game=pd.read_csv('tic-tac-toe.csv')
game.head()

TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True

In [59]:
#Checking data types and statistics
display(game.info())
display(game.describe().T)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB


None

count unique   top freq
TL      958      3     x  418
TM      958      3     x  378
TR      958      3     x  418
ML      958      3     x  378
MM      958      3     x  458
MR      958      3     x  378
BL      958      3     x  418
BM      958      3     x  378
BR      958      3     x  418
class   958      2  True  626

In [60]:
#Convert the categorical values to numeric in all columns
game = pd.get_dummies(game)

In [61]:
game.head()

class   TL_b   TL_o  TL_x   TM_b   TM_o  TM_x   TR_b   TR_o  TR_x  ...  \
0   True  False  False  True  False  False  True  False  False  True  ...   
1   True  False  False  True  False  False  True  False  False  True  ...   
2   True  False  False  True  False  False  True  False  False  True  ...   
3   True  False  False  True  False  False  True  False  False  True  ...   
4   True  False  False  True  False  False  True  False  False  True  ...   

    MR_x   BL_b   BL_o   BL_x   BM_b   BM_o   BM_x   BR_b   BR_o   BR_x  
0  False  False  False   True  False   True  False  False   True  False  
1  False  False   True  False  False  False   True  False   True  False  
2  False  False   True  False  False   True  False  False  False   True  
3  False  False   True  False   True  False  False   True  False  False  
4  False   True  False  False  False   True  False   True  False  False  

[5 rows x 28 columns]

In [62]:
#Separate the inputs and output
x = game.drop(columns='class')
y = game['class'].astype(int)


In [63]:
#Normalize the input data
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
x_normalized = scaler.fit_transform(x)

display(pd.DataFrame(x_normalized, columns=x.columns).head())


TL_b      TL_o      TL_x      TM_b      TM_o      TM_x     TR_b  \
0 -0.52177 -0.733294  1.136603 -0.594228 -0.724898  1.238706 -0.52177   
1 -0.52177 -0.733294  1.136603 -0.594228 -0.724898  1.238706 -0.52177   
2 -0.52177 -0.733294  1.136603 -0.594228 -0.724898  1.238706 -0.52177   
3 -0.52177 -0.733294  1.136603 -0.594228 -0.724898  1.238706 -0.52177   
4 -0.52177 -0.733294  1.136603 -0.594228 -0.724898  1.238706 -0.52177   

       TR_o      TR_x      ML_b  ...      MR_x      BL_b      BL_o      BL_x  \
0 -0.733294  1.136603 -0.594228  ... -0.807294 -0.521770 -0.733294  1.136603   
1 -0.733294  1.136603 -0.594228  ... -0.807294 -0.521770  1.363709 -0.879815   
2 -0.733294  1.136603 -0.594228  ... -0.807294 -0.521770  1.363709 -0.879815   
3 -0.733294  1.136603 -0.594228  ... -0.807294 -0.521770  1.363709 -0.879815   
4 -0.733294  1.136603 -0.594228  ... -0.807294  1.916552 -0.733294 -0.879815   

       BM_b      BM_o      BM_x      BR_b      BR_o      BR_x  
0 -0.594228  1.379504 -0.807294 -0.521770  1.363709 -0.879815  
1 -0.594228 -0.724898  1.238706 -0.521770  1.363709 -0.879815  
2 -0.594228  1.379504 -0.807294 -0.521770 -0.733294  1.136603  
3  1.682855 -0.724898 -0.807294  1.916552 -0.733294 -0.879815  
4 -0.594228  1.379504 -0.807294  1.916552 -0.733294 -0.879815  

[5 rows x 27 columns]

## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [64]:
# your code here
# Split the training and test data.
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_normalized, y, random_state=42, test_size=0.2)


In [65]:
# Create a `Sequential` model.
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential()

In [66]:
# Add several layers to your model. 
# Make sure you use ReLU as the activation function for the middle layers. 
# Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.

model.add(Dense(64, activation='relu', input_shape=(x_train.shape[1],)))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

C:\Users\aanas\AppData\Roaming\Python\Python311\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [67]:
# Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. 
# For metrics, use `accuracy` for now.

#Replaced "sparse_categorical_crossentropy" with "binary_crossentropy" because I kept getting an error fitting the training data
#When searching for the reason, I found that binary is more accurate for binary data (which is what I have)

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])


In [68]:
# Fit the training data.
history = model.fit(x_train, y_train, epochs=50, batch_size=32, validation_split=0.2)

Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.6288 - loss: 0.6751 - val_accuracy: 0.7013 - val_loss: 0.5705
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7152 - loss: 0.5385 - val_accuracy: 0.7338 - val_loss: 0.5338
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8091 - loss: 0.4672 - val_accuracy: 0.7338 - val_loss: 0.5186
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8270 - loss: 0.4283 - val_accuracy: 0.7727 - val_loss: 0.4911
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8157 - loss: 0.4256 - val_accuracy: 0.7857 - val_loss: 0.4670
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8853 - loss: 0.3297 - val_accuracy: 0.7857 - val_loss: 0.4447
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8920 - loss: 0.3197 - val_accuracy: 0.8247 - val_loss: 0.4079
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8983 - loss: 0.2921 - val_accuracy: 0.8247 - val_loss

In [69]:
# Evaluate your neural network model with the test data.
test_loss, test_accuracy = model.evaluate(x_test, y_test)

print (f'Test loss: {test_loss}')
print (f'Test accuracy: {test_accuracy}')

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.9448 - loss: 0.1174 
Test loss: 0.08930452913045883
Test accuracy: 0.9635416865348816


In [70]:
# Save your model as `tic-tac-toe.model`.
model.save('tic-tac-toe.keras')

## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [71]:
# your code here
from tensorflow.keras.models import load_model

model = load_model('tic-tac-toe.keras')

In [72]:
#Making predictions on random rows
import numpy as np

random_indices = np.random.choice(x_test.shape[0], size=5, replace= False)
x_sample = x_test[random_indices]
y_sample = y_test.iloc[random_indices]

predictions = model.predict(x_sample)
predictions = (predictions > 0.5).astype(int) #predictions of player X wins

#displaying predictions and actual values
for i in range(len(predictions)):
    print(f'Predicted: {predictions[i][0]}, Actual: {y_sample.iloc[i]}')

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step
Predicted: 1, Actual: 1
Predicted: 1, Actual: 0
Predicted: 1, Actual: 1
Predicted: 1, Actual: 1
Predicted: 1, Actual: 1


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [82]:
# your code here
from tensorflow.keras.optimizers import Adam 
from tensorflow.keras.callbacks import EarlyStopping

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(x_normalized, y, test_size=0.2, random_state=42)

# Create a Sequential model
model = Sequential()

# Adding more layers to the previous model. 
model.add(Dense(128, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dense(128, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

# Create a custom Adam optimizer with a specified learning rate
learning_rate = 0.01  
optimizer = Adam(learning_rate= learning_rate)

# Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. 
# For metrics, use `accuracy` for now.
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])

# Early stopping to prevent overfitting
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

# Fit the training data.
epochs = 100
history = model.fit(x_train, y_train, epochs=epochs, batch_size=32, validation_split=0.2)

# Evaluate your neural network model with the test data.
test_loss, test_accuracy = model.evaluate(x_test, y_test)

print (f'Test loss: {test_loss}')
print (f'Test accuracy: {test_accuracy}')


Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 3s 14ms/step - accuracy: 0.6294 - loss: 0.6291 - val_accuracy: 0.7662 - val_loss: 0.5496
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8206 - loss: 0.4137 - val_accuracy: 0.8377 - val_loss: 0.3086
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9300 - loss: 0.1814 - val_accuracy: 0.9091 - val_loss: 0.2640
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9679 - loss: 0.1104 - val_accuracy: 0.8896 - val_loss: 0.4249
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9523 - loss: 0.1492 - val_accuracy: 0.9351 - val_loss: 0.2981
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9851 - loss: 0.0608 - val_accuracy: 0.9675 - val_loss: 0.0758
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9827 - loss: 0.0389 - val_accuracy: 0.9091 - val_loss: 0.3650
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9733 - loss: 0.1129 - val_accuracy: 0.9740 - 

**Which approach(es) did you find helpful to improve your model performance?**

# your answer here
The 1st model had:
    4 layers
    optimizer with default learning rate 
    epochs = 50
And returned:
    Test loss: 0.08930452913045883
    Test accuracy: 0.9635416865348816


On 1st test to improve the model, I had: 
    6 layers (+2 layers)
    optimizer learning rate = 0.001
    epochs = 100
And returned:
    Test loss: 0.12479975819587708
    Test accuracy: 0.953125
DECREASED PERFORMANCE

On 2nd test to improve the model, I updated learning rate only: 
    6 layers 
    **optimizer learning rate** = 0.0001
    epochs = 100
And returned:
    Test loss: 0.11020072549581528
    Test accuracy: 0.953125
Accuracy remain the same, test loss decreased

On 3rd test to improve the model, I updated learning rate only: 
    6 layers 
    **optimizer learning rate** = 0.01
    epochs = 100
And returned:
    Test loss: 0.10970013588666916
    Test accuracy: 0.9791666865348816
Improved the accuracy and the test loss

On 4th test to improve the model, I updated epochs only: 
    6 layers 
    optimizer learning rate = 0.01
    **epochs** = 150
And returned:
    Test loss: 0.08076969534158707
    Test accuracy: 0.9791666865348816
Increased the accuracy and the test loss


Best would be to pick the hyperparameters values of the 4th option.
